packages/tpetra/core/test/MultiVector/MultiVector_MicroBenchmark.cpp

```diff
@@ -0,0 +1,149 @@
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ************************************************************************
+// @HEADER
+*/
+
+#include "Tpetra_TestingUtilities.hpp"
+#include "Tpetra_Map.hpp"
+#include "Tpetra_MultiVector.hpp"
+#include "Teuchos_ScalarTraits.hpp"
+#include "Kokkos_ArithTraits.hpp"
+#include "Tpetra_Details_Profiling.hpp"
+#include "KokkosBlas.hpp"
+
+namespace { // (anonymous)
+
+  using Tpetra::TestingUtilities::getDefaultComm;
+  using Teuchos::RCP;
+  using std::endl;
+
+  //
+  // UNIT TESTS
+  //
+
+  ////
+  TEUCHOS_UNIT_TEST_TEMPLATE_4_DECL( MultiVector, MicroBenchmark_Update, SC, LO, GO, NT )
+  {
+    typedef Tpetra::Map<LO, GO, NT> map_type;
+    typedef Tpetra::MultiVector<SC, LO, GO, NT> MV;
+    typedef Teuchos::ScalarTraits<SC> STS;
+    
+    SC ONE = STS::one();
+    SC ZERO = STS::zero();
+
+    // Problem parameters
+    size_t vector_size = 125000;
+    size_t num_repeats = 2000;
+
+    auto comm = getDefaultComm ();
+    const int numProcs = comm->getSize ();
+    if(numProcs != 1) return;
+
+    const LO lclNumRows = vector_size;
+    const GO indexBase = 0;
+    RCP<const map_type> map (new map_type (lclNumRows, lclNumRows,
+                                           indexBase, comm));
+    const LO numVecs = 1;  // FIXME: This needs to be 1 for the lambda loop to work
+    MV X (map, numVecs), Y(map,numVecs);
+    X.putScalar(ZERO);
+    Y.putScalar(ONE);
+
+    Kokkos::fence();
+
+
+    typedef typename MV::dual_view_type::t_dev dev_view_type;
+    //    typedef typename MV::dual_view_type::t_host host_view_type;
+    typedef typename dev_view_type::memory_space cur_memory_space;
+    typedef typename dev_view_type::execution_space cur_exec_space;
+
+    // Run Tpetra Update Loop
+    {
+      ::Tpetra::Details::ProfilingRegion region ("Tpetra Update Loop");
+      for(size_t i=0; i<num_repeats; i++) {
+        X.update(ONE,Y,ONE);
+      }
```

Comment:
You actually want the fence here, within the timing scope but outside (and after) all the repeats.  This will ensure accurate timings, at least if we can ever turn off `CUDA_LAUNCH_BLOCKING`.